// @ionic-selectable/angular/esm2015/utils/ionic-selectable-value-accessor.js
import { Directive, ElementRef, HostListener, Injector } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { ValueAccessor } from './value-accessor';
export class IonicSelectableValueAccessor extends ValueAccessor {
    constructor(injector, el) {
        super(injector, el);
    }
    _handleChangeEvent(el) {
        this.handleChangeEvent(el, el.value);
    }
}
IonicSelectableValueAccessor.decorators = [
    { type: Directive, args: [{
                /* tslint:disable-next-line:directive-selector */
                selector: 'ionic-selectable',
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: IonicSelectableValueAccessor,
                        multi: true
                    }
                ]
            },] }
];
IonicSelectableValueAccessor.ctorParameters = () => [
    { type: Injector },
    { type: ElementRef }
];
IonicSelectableValueAccessor.propDecorators = {
    _handleChangeEvent: [{ type: HostListener, args: ['changed', ['$event.target'],] }]
};
