// @awesome-cordova-plugins/document-viewer/__ivy_ngcc__/ngx/index.js
import { __extends } from "tslib";
import { AwesomeCordovaNativePlugin, cordova } from '@awesome-cordova-plugins/core';
import { Injectable } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
var DocumentViewer = /** @class */ (function (_super) {
    __extends(DocumentViewer, _super);
    function DocumentViewer() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DocumentViewer.prototype.getSupportInfo = function () { return cordova(this, "getSupportInfo", {}, arguments); };
    DocumentViewer.prototype.canViewDocument = function (url, contentType, options, onPossible, onMissingApp, onImpossible, onError) { return cordova(this, "canViewDocument", { "sync": true }, arguments); };
    DocumentViewer.prototype.viewDocument = function (url, contentType, options, onShow, onClose, onMissingApp, onError) { return cordova(this, "viewDocument", { "sync": true }, arguments); };
    DocumentViewer.pluginName = "Document Viewer";
    DocumentViewer.plugin = "cordova-plugin-document-viewer";
    DocumentViewer.pluginRef = "SitewaertsDocumentViewer";
    DocumentViewer.repo = "https://github.com/sitewaerts/cordova-plugin-document-viewer";
    DocumentViewer.platforms = ["Android", "iOS", "Windows"];
DocumentViewer.ɵfac = /*@__PURE__*/ function () { var ɵDocumentViewer_BaseFactory; return function DocumentViewer_Factory(t) { return (ɵDocumentViewer_BaseFactory || (ɵDocumentViewer_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(DocumentViewer)))(t || DocumentViewer); }; }();
DocumentViewer.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: DocumentViewer, factory: function (t) { return DocumentViewer.ɵfac(t); } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DocumentViewer, [{
        type: Injectable
    }], null, null); })();
    return DocumentViewer;
}(AwesomeCordovaNativePlugin));
export { DocumentViewer };
