// @ionic-selectable/core/dist/esm/ionic-selectable.interfaces.component-dca593f8.js
class IonicSelectableEvent {
  constructor(value, component) {
    this.value = value;
    this.component = component;
  }
}
class IonicSelectableInfiniteScrolledEvent extends IonicSelectableEvent {
}
class IonicSelectableSearchingEvent extends IonicSelectableEvent {
}
class IonicSelectableSearchFailedEvent extends IonicSelectableEvent {
}
class IonicSelectableSearchSuccessedEvent extends IonicSelectableEvent {
}
class IonicSelectableItemAddingEvent extends IonicSelectableEvent {
}
class IonicSelectableClearedEvent extends IonicSelectableEvent {
}
class IonicSelectableChangedEvent extends IonicSelectableEvent {
}
class IonicSelectableItemsChangedEvent extends IonicSelectableEvent {
}
class IonicSelectableSelectedEvent extends IonicSelectableEvent {
  constructor(value, isSelected, component) {
    super(value, component);
    this.isSelected = isSelected;
  }
}
class IonicSelectableClosedEvent extends IonicSelectableEvent {
}
class IonicSelectableOpenedEvent extends IonicSelectableEvent {
}
class IonicSelectableFocusedEvent extends IonicSelectableEvent {
}
class IonicSelectableBlurredEvent extends IonicSelectableEvent {
}

export { IonicSelectableEvent as I, IonicSelectableInfiniteScrolledEvent as a, IonicSelectableSearchingEvent as b, IonicSelectableSearchFailedEvent as c, IonicSelectableSearchSuccessedEvent as d, IonicSelectableItemAddingEvent as e, IonicSelectableClearedEvent as f, IonicSelectableChangedEvent as g, IonicSelectableItemsChangedEvent as h, IonicSelectableSelectedEvent as i, IonicSelectableClosedEvent as j, IonicSelectableOpenedEvent as k, IonicSelectableFocusedEvent as l, IonicSelectableBlurredEvent as m };
