// @ionic-selectable/core/dist/cjs/ionic-selectable.interfaces.component-f6bd7825.js
'use strict';

class IonicSelectableEvent {
  constructor(value, component) {
    this.value = value;
    this.component = component;
  }
}
class IonicSelectableInfiniteScrolledEvent extends IonicSelectableEvent {
}
class IonicSelectableSearchingEvent extends IonicSelectableEvent {
}
class IonicSelectableSearchFailedEvent extends IonicSelectableEvent {
}
class IonicSelectableSearchSuccessedEvent extends IonicSelectableEvent {
}
class IonicSelectableItemAddingEvent extends IonicSelectableEvent {
}
class IonicSelectableClearedEvent extends IonicSelectableEvent {
}
class IonicSelectableChangedEvent extends IonicSelectableEvent {
}
class IonicSelectableItemsChangedEvent extends IonicSelectableEvent {
}
class IonicSelectableSelectedEvent extends IonicSelectableEvent {
  constructor(value, isSelected, component) {
    super(value, component);
    this.isSelected = isSelected;
  }
}
class IonicSelectableClosedEvent extends IonicSelectableEvent {
}
class IonicSelectableOpenedEvent extends IonicSelectableEvent {
}
class IonicSelectableFocusedEvent extends IonicSelectableEvent {
}
class IonicSelectableBlurredEvent extends IonicSelectableEvent {
}

exports.IonicSelectableBlurredEvent = IonicSelectableBlurredEvent;
exports.IonicSelectableChangedEvent = IonicSelectableChangedEvent;
exports.IonicSelectableClearedEvent = IonicSelectableClearedEvent;
exports.IonicSelectableClosedEvent = IonicSelectableClosedEvent;
exports.IonicSelectableEvent = IonicSelectableEvent;
exports.IonicSelectableFocusedEvent = IonicSelectableFocusedEvent;
exports.IonicSelectableInfiniteScrolledEvent = IonicSelectableInfiniteScrolledEvent;
exports.IonicSelectableItemAddingEvent = IonicSelectableItemAddingEvent;
exports.IonicSelectableItemsChangedEvent = IonicSelectableItemsChangedEvent;
exports.IonicSelectableOpenedEvent = IonicSelectableOpenedEvent;
exports.IonicSelectableSearchFailedEvent = IonicSelectableSearchFailedEvent;
exports.IonicSelectableSearchSuccessedEvent = IonicSelectableSearchSuccessedEvent;
exports.IonicSelectableSearchingEvent = IonicSelectableSearchingEvent;
exports.IonicSelectableSelectedEvent = IonicSelectableSelectedEvent;
