// @ionic-selectable/core/dist/collection/components/ionic-selectable/ionic-selectable.interfaces.component.js
export class IonicSelectableEvent {
  constructor(value, component) {
    this.value = value;
    this.component = component;
  }
}
export class IonicSelectableInfiniteScrolledEvent extends IonicSelectableEvent {
}
export class IonicSelectableSearchingEvent extends IonicSelectableEvent {
}
export class IonicSelectableSearchFailedEvent extends IonicSelectableEvent {
}
export class IonicSelectableSearchSuccessedEvent extends IonicSelectableEvent {
}
export class IonicSelectableItemAddingEvent extends IonicSelectableEvent {
}
export class IonicSelectableClearedEvent extends IonicSelectableEvent {
}
export class IonicSelectableChangedEvent extends IonicSelectableEvent {
}
export class IonicSelectableItemsChangedEvent extends IonicSelectableEvent {
}
export class IonicSelectableSelectedEvent extends IonicSelectableEvent {
  constructor(value, isSelected, component) {
    super(value, component);
    this.isSelected = isSelected;
  }
}
export class IonicSelectableClosedEvent extends IonicSelectableEvent {
}
export class IonicSelectableOpenedEvent extends IonicSelectableEvent {
}
export class IonicSelectableFocusedEvent extends IonicSelectableEvent {
}
export class IonicSelectableBlurredEvent extends IonicSelectableEvent {
}
